File: apps/mobile/lib/analysis/domain/wealth_target_allocation.dart
import '../../core/decimals.dart';
import '../../dashboard/logic/dashboard_aggregate.dart';
import 'wealth_analysis.dart';

class WealthTarget {
  const WealthTarget({required this.assetClass, required this.percentage});

  final AssetGroup assetClass;
  final String percentage;
}

class WealthTargetPlan {
  const WealthTargetPlan({
    required this.targets,
    required this.tolerancePercentage,
  });

  final List<WealthTarget> targets;
  final String tolerancePercentage;
}

class WealthTargetInputSummary {
  const WealthTargetInputSummary({
    required this.total,
    required this.isValid,
    required this.targets,
  });

  final String total;
  final bool isValid;
  final List<WealthTarget>? targets;
}

class WealthToleranceInputSummary {
  const WealthToleranceInputSummary({
    required this.isValid,
    required this.tolerancePercentage,
  });

  final bool isValid;
  final String? tolerancePercentage;
}

enum WealthTargetDriftStatus { within, above, below }

class WealthTargetDriftRow {
  const WealthTargetDriftRow({
    required this.assetClass,
    required this.currentPercentage,
    required this.targetPercentage,
    required this.lowerBoundPercentage,
    required this.upperBoundPercentage,
    required this.gapPercentage,
    required this.monetaryGap,
    required this.status,
  });

  final WealthAssetClass assetClass;
  final String currentPercentage;
  final String targetPercentage;
  final String lowerBoundPercentage;
  final String upperBoundPercentage;
  final String gapPercentage;
  final String monetaryGap;
  final WealthTargetDriftStatus status;
}

enum WealthTargetComparisonStatus { notConfigured, unavailable, available }

enum WealthTargetUnavailableReason {
  invalidTargets,
  invalidTolerance,
  valuationIncomplete,
  participatingValueUnavailable,
  zeroComparisonBase,
}

class WealthTargetComparison {
  const WealthTargetComparison._({
    required this.status,
    this.reason,
    this.comparisonBase,
    this.rows = const [],
    this.largestDeviation,
  });

  const WealthTargetComparison.notConfigured()
    : this._(status: WealthTargetComparisonStatus.notConfigured);

  const WealthTargetComparison.unavailable(WealthTargetUnavailableReason reason)
    : this._(status: WealthTargetComparisonStatus.unavailable, reason: reason);

  const WealthTargetComparison.available({
    required String comparisonBase,
    required List<WealthTargetDriftRow> rows,
    required WealthTargetDriftRow? largestDeviation,
  }) : this._(
         status: WealthTargetComparisonStatus.available,
         comparisonBase: comparisonBase,
         rows: rows,
         largestDeviation: largestDeviation,
       );

  final WealthTargetComparisonStatus status;
  final WealthTargetUnavailableReason? reason;
  final String? comparisonBase;
  final List<WealthTargetDriftRow> rows;
  final WealthTargetDriftRow? largestDeviation;
}

final _percentagePattern = RegExp(r'^\d+(?:\.\d{1,6})?$');

String? _validPercentage(String value) {
  if (!_percentagePattern.hasMatch(value)) return null;
  final normalized = D.normalize(value);
  if (normalized == null ||
      (D.compare(normalized, '0') ?? 0) < 0 ||
      (D.compare(normalized, '100') ?? 0) > 0) {
    return null;
  }
  return normalized;
}

WealthToleranceInputSummary summarizeWealthToleranceInput(String input) {
  final tolerance = _validPercentage(input.isEmpty ? '0' : input);
  return WealthToleranceInputSummary(
    isValid: tolerance != null,
    tolerancePercentage: tolerance,
  );
}

WealthTargetInputSummary summarizeWealthTargetInputs(
  Map<AssetGroup, String> inputs,
) {
  var total = '0';
  var valid = true;
  final targets = <WealthTarget>[];
  for (final assetClass in wealthAssetGroups) {
    final normalized = _validPercentage(inputs[assetClass] ?? '');
    if (normalized == null) {
      valid = false;
      continue;
    }
    total = D.add(total, normalized)!;
    targets.add(WealthTarget(assetClass: assetClass, percentage: normalized));
  }
  final isValid =
      valid &&
      targets.length == wealthAssetGroups.length &&
      D.compare(total, '100') == 0;
  return WealthTargetInputSummary(
    total: total,
    isValid: isValid,
    targets: isValid ? targets : null,
  );
}

bool _targetsAreValid(List<WealthTarget> targets) {
  if (targets.length != wealthAssetGroups.length) return false;
  return summarizeWealthTargetInputs({
    for (final target in targets) target.assetClass: target.percentage,
  }).isValid;
}

String _absolute(String value) =>
    (D.compare(value, '0') ?? 0) < 0 ? D.subtract('0', value)! : value;

WealthTargetComparison calculateWealthTargetComparison({
  required DashboardAggregate aggregate,
  required WealthAnalysisEvidence evidence,
  required List<WealthTarget> targets,
  String? tolerancePercentage,
}) {
  if (targets.isEmpty) return const WealthTargetComparison.notConfigured();
  if (!_targetsAreValid(targets)) {
    return const WealthTargetComparison.unavailable(
      WealthTargetUnavailableReason.invalidTargets,
    );
  }
  final tolerance = _validPercentage(tolerancePercentage ?? '0');
  if (tolerance == null) {
    return const WealthTargetComparison.unavailable(
      WealthTargetUnavailableReason.invalidTolerance,
    );
  }
  if (aggregate.status != AggregateStatus.complete) {
    return const WealthTargetComparison.unavailable(
      WealthTargetUnavailableReason.valuationIncomplete,
    );
  }

  final participating = targets
      .where((target) => (D.compare(target.percentage, '0') ?? 0) > 0)
      .toList(growable: false);
  final byGroup = {
    for (final assetClass in evidence.assetClasses)
      assetClass.group: assetClass,
  };
  final participatingClasses = participating
      .map((target) => (target: target, asset: byGroup[target.assetClass]))
      .toList(growable: false);
  if (participatingClasses.any((item) => item.asset?.value == null)) {
    return const WealthTargetComparison.unavailable(
      WealthTargetUnavailableReason.participatingValueUnavailable,
    );
  }
  final comparisonBase = D.sum(
    participatingClasses.map((item) => item.asset!.value),
  );
  if (!D.isPositive(comparisonBase)) {
    return const WealthTargetComparison.unavailable(
      WealthTargetUnavailableReason.zeroComparisonBase,
    );
  }

  final rows = <WealthTargetDriftRow>[];
  for (final item in participatingClasses) {
    final currentRatio = D.divide(
      item.asset!.value,
      comparisonBase,
      scale: 12,
    )!;
    final targetRatio = D.divide(item.target.percentage, '100', scale: 12)!;
    final currentPercentage = D.multiply(currentRatio, '100')!;
    final targetImpliedValue = D.multiply(comparisonBase, targetRatio)!;
    final gapPercentage = D.subtract(
      currentPercentage,
      item.target.percentage,
    )!;
    final monetaryGap = D.subtract(item.asset!.value, targetImpliedValue)!;
    final rawLower = D.subtract(item.target.percentage, tolerance)!;
    final rawUpper = D.add(item.target.percentage, tolerance)!;
    final lower = (D.compare(rawLower, '0') ?? 0) < 0 ? '0' : rawLower;
    final upper = (D.compare(rawUpper, '100') ?? 0) > 0 ? '100' : rawUpper;
    final status = (D.compare(currentPercentage, lower) ?? 0) < 0
        ? WealthTargetDriftStatus.below
        : (D.compare(currentPercentage, upper) ?? 0) > 0
        ? WealthTargetDriftStatus.above
        : WealthTargetDriftStatus.within;
    rows.add(
      WealthTargetDriftRow(
        assetClass: item.asset!,
        currentPercentage: currentPercentage,
        targetPercentage: item.target.percentage,
        lowerBoundPercentage: lower,
        upperBoundPercentage: upper,
        gapPercentage: gapPercentage,
        monetaryGap: monetaryGap,
        status: status,
      ),
    );
  }

  WealthTargetDriftRow? largest;
  for (final row in rows.where(
    (row) => row.status != WealthTargetDriftStatus.within,
  )) {
    if (largest == null ||
        (D.compare(
                  _absolute(row.gapPercentage),
                  _absolute(largest.gapPercentage),
                ) ??
                0) >
            0) {
      largest = row;
    }
  }
  return WealthTargetComparison.available(
    comparisonBase: comparisonBase!,
    rows: rows,
    largestDeviation: largest,
  );
}

List<WealthAssetClass> getExcludedValuedTargetClasses(
  WealthAnalysisEvidence evidence,
  List<WealthTarget> targets,
) {
  final byGroup = {
    for (final target in targets) target.assetClass: target.percentage,
  };
  return evidence.assetClasses
      .where(
        (asset) =>
            D.compare(byGroup[asset.group], '0') == 0 &&
            D.isPositive(asset.value),
      )
      .toList(growable: false);
}

File: apps/web/src/features/analysis/domain/wealth-target-allocation.ts
import type {
  DashboardAggregate,
  DashboardAssetGroup,
} from "@/features/dashboard/services/dashboard-aggregate.service"
import type { Decimal } from "@/lib/supabase/types"
import {
  addDecimals,
  compareDecimals,
  divideDecimals,
  multiplyDecimals,
  normalizeDecimal,
  subtractDecimals,
} from "@/lib/financial-calculations/decimal"

import type {
  WealthAnalysisEvidence,
  WealthAssetClass,
} from "../utils/wealth-analysis"

export type WealthTarget = {
  assetClass: DashboardAssetGroup
  percentage: Decimal
}

export type WealthTargetPlan = {
  targets: WealthTarget[]
  tolerancePercentage: Decimal
}

export type WealthTargetInputSummary = {
  total: Decimal
  isValid: boolean
  targets: WealthTarget[] | null
}

export type WealthTargetDriftStatus = "within" | "above" | "below"

export type WealthTargetDriftRow = {
  assetClass: WealthAssetClass
  currentPercentage: Decimal
  targetPercentage: Decimal
  lowerBoundPercentage: Decimal
  upperBoundPercentage: Decimal
  gapPercentage: Decimal
  monetaryGap: Decimal
  status: WealthTargetDriftStatus
}

export type WealthTargetComparison =
  | { status: "not-configured" }
  | {
      status: "unavailable"
      reason:
        | "invalid-targets"
        | "invalid-tolerance"
        | "valuation-incomplete"
        | "participating-value-unavailable"
        | "zero-comparison-base"
    }
  | {
      status: "available"
      comparisonBase: Decimal
      rows: WealthTargetDriftRow[]
      largestDeviation: WealthTargetDriftRow | null
    }

function addOrThrow(left: Decimal, right: Decimal): Decimal {
  const result = addDecimals(left, right)
  if (result === null) throw new Error("Invalid target decimal")
  return result
}

function absoluteDecimal(value: Decimal): Decimal | null {
  const comparison = compareDecimals(value, "0")
  if (comparison === null) return null
  return comparison === -1 ? subtractDecimals("0", value) : value
}

function validPercentage(value: unknown): Decimal | null {
  if (typeof value !== "string") return null
  if (!/^[0-9]+(?:\.[0-9]{1,6})?$/.test(value)) return null
  const normalized = normalizeDecimal(value)
  if (
    normalized === null ||
    compareDecimals(normalized, "0") === -1 ||
    compareDecimals(normalized, "100") === 1
  ) {
    return null
  }
  return normalized
}

export function summarizeWealthToleranceInput(input: string): {
  isValid: boolean
  tolerancePercentage: Decimal | null
} {
  const tolerancePercentage = validPercentage(input === "" ? "0" : input)
  return {
    isValid: tolerancePercentage !== null,
    tolerancePercentage,
  }
}

export function summarizeWealthTargetInputs(
  inputs: Readonly<Record<string, string>>,
  supportedAssetClasses: readonly DashboardAssetGroup[]
): WealthTargetInputSummary {
  let total: Decimal = "0"
  let valid = true
  const targets: WealthTarget[] = []

  for (const assetClass of supportedAssetClasses) {
    const input = inputs[assetClass] ?? ""
    const normalized = validPercentage(input)
    if (normalized === null) {
      valid = false
      continue
    }
    total = addOrThrow(total, normalized)
    targets.push({ assetClass, percentage: normalized })
  }

  const isValid =
    valid &&
    targets.length === supportedAssetClasses.length &&
    compareDecimals(total, "100") === 0

  return { total, isValid, targets: isValid ? targets : null }
}

function targetsAreValid(
  targets: readonly WealthTarget[],
  supportedAssetClasses: readonly DashboardAssetGroup[]
) {
  if (targets.length !== supportedAssetClasses.length) return false
  const inputs = Object.fromEntries(
    targets.map(({ assetClass, percentage }) => [assetClass, percentage])
  )
  return summarizeWealthTargetInputs(inputs, supportedAssetClasses).isValid
}

export function calculateWealthTargetComparison(
  aggregate: DashboardAggregate,
  evidence: WealthAnalysisEvidence,
  targets: readonly WealthTarget[],
  tolerancePercentage: Decimal | null
): WealthTargetComparison {
  if (targets.length === 0) return { status: "not-configured" }
  const supportedAssetClasses = evidence.assetClasses.map(({ group }) => group)
  if (!targetsAreValid(targets, supportedAssetClasses)) {
    return { status: "unavailable", reason: "invalid-targets" }
  }
  const normalizedTolerance = validPercentage(tolerancePercentage ?? "0")
  if (normalizedTolerance === null) {
    return { status: "unavailable", reason: "invalid-tolerance" }
  }
  if (aggregate.status !== "complete") {
    return { status: "unavailable", reason: "valuation-incomplete" }
  }

  const participating = targets.filter(
    ({ percentage }) => compareDecimals(percentage, "0") === 1
  )
  const classesByGroup = new Map(
    evidence.assetClasses.map((assetClass) => [assetClass.group, assetClass])
  )
  const participatingClasses = participating.map((target) => ({
    target,
    assetClass: classesByGroup.get(target.assetClass) ?? null,
  }))
  if (
    participatingClasses.some(
      ({ assetClass }) => assetClass === null || assetClass.value === null
    )
  ) {
    return {
      status: "unavailable",
      reason: "participating-value-unavailable",
    }
  }

  const comparisonBase = participatingClasses.reduce<Decimal>(
    (total, { assetClass }) => addOrThrow(total, assetClass!.value!),
    "0"
  )
  if (compareDecimals(comparisonBase, "0") !== 1) {
    return { status: "unavailable", reason: "zero-comparison-base" }
  }

  const rows = participatingClasses.map<WealthTargetDriftRow>(
    ({ target, assetClass }) => {
      const currentRatio = divideDecimals(
        assetClass!.value!,
        comparisonBase,
        12
      )
      const targetRatio = divideDecimals(target.percentage, "100", 12)
      if (currentRatio === null || targetRatio === null) {
        throw new Error("Unable to calculate target allocation ratio")
      }
      const currentPercentage = multiplyDecimals(currentRatio, "100")
      const targetImpliedValue = multiplyDecimals(comparisonBase, targetRatio)
      if (currentPercentage === null || targetImpliedValue === null) {
        throw new Error("Unable to calculate target allocation values")
      }
      const gapPercentage = subtractDecimals(
        currentPercentage,
        target.percentage
      )
      const monetaryGap = subtractDecimals(
        assetClass!.value!,
        targetImpliedValue
      )
      if (gapPercentage === null || monetaryGap === null) {
        throw new Error("Unable to calculate target allocation gaps")
      }
      const unclampedLower = subtractDecimals(
        target.percentage,
        normalizedTolerance
      )
      const unclampedUpper = addDecimals(target.percentage, normalizedTolerance)
      if (unclampedLower === null || unclampedUpper === null) {
        throw new Error("Unable to calculate target allocation range")
      }
      const lowerBoundPercentage =
        compareDecimals(unclampedLower, "0") === -1 ? "0" : unclampedLower
      const upperBoundPercentage =
        compareDecimals(unclampedUpper, "100") === 1 ? "100" : unclampedUpper
      const belowLower = compareDecimals(
        currentPercentage,
        lowerBoundPercentage
      )
      const aboveUpper = compareDecimals(
        currentPercentage,
        upperBoundPercentage
      )
      if (belowLower === null || aboveUpper === null) {
        throw new Error("Unable to classify target allocation range")
      }
      return {
        assetClass: assetClass!,
        currentPercentage,
        targetPercentage: target.percentage,
        lowerBoundPercentage,
        upperBoundPercentage,
        gapPercentage,
        monetaryGap,
        status:
          belowLower === -1 ? "below" : aboveUpper === 1 ? "above" : "within",
      }
    }
  )

  const largestDeviation = rows
    .filter(({ status }) => status !== "within")
    .reduce<WealthTargetDriftRow | null>((largest, row) => {
      if (!largest) return row
      const currentAbsolute = absoluteDecimal(row.gapPercentage)
      const largestAbsolute = absoluteDecimal(largest.gapPercentage)
      if (currentAbsolute === null || largestAbsolute === null) return largest
      return compareDecimals(currentAbsolute, largestAbsolute) === 1
        ? row
        : largest
    }, null)

  return {
    status: "available",
    comparisonBase,
    rows,
    largestDeviation,
  }
}

export function getExcludedValuedTargetClasses(
  evidence: WealthAnalysisEvidence,
  targets: readonly WealthTarget[]
): WealthAssetClass[] {
  const targetsByClass = new Map(
    targets.map(({ assetClass, percentage }) => [assetClass, percentage])
  )
  return evidence.assetClasses.filter((assetClass) => {
    const target = targetsByClass.get(assetClass.group)
    return (
      target !== undefined &&
      compareDecimals(target, "0") === 0 &&
      assetClass.value !== null &&
      compareDecimals(assetClass.value, "0") === 1
    )
  })
}
